core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala

```diff
@@ -0,0 +1,369 @@
+        }
+      })
+    })
+    (1 to numReplicaFetchers).foreach(i => {
+      futures += executorService.submit(new Runnable {
+        override def run(): Unit = {
+          try {
+            updateFollowerFetchState(partition, i, totalRecords, followerQueues(i -1))
+          } catch {
+            case e: Throwable =>
+              PartitionLockTest.this.error("Exception during updateFollowerFetchState", e)
+              throw e
+          }
+        }
+      })
+    })
+    futures.foreach(_.get())
+  }
+
+  private def scheduleShrinkIsr(activeFlag: AtomicBoolean, mockTimeSleepMs: Long): Future[_] = {
+    executorService.submit(new Runnable {
+      override def run(): Unit = {
+        while (activeFlag.get) {
+          if (mockTimeSleepMs > 0)
+            mockTime.sleep(mockTimeSleepMs)
+          partition.maybeShrinkIsr()
+          Thread.sleep(1) // just to avoid tight loop
+        }
+      }
+    })
+  }
+
+  private def setupPartitionWithMocks(logManager: LogManager, logConfig: LogConfig): Partition = {
+    val leaderEpoch = 1
+    val brokerId = 0
+    val topicPartition = new TopicPartition("test-topic", 0)
+    val stateStore: PartitionStateStore = mock(classOf[PartitionStateStore])
+    val delayedOperations: DelayedOperations = mock(classOf[DelayedOperations])
+    val metadataCache: MetadataCache = mock(classOf[MetadataCache])
+    val offsetCheckpoints: OffsetCheckpoints = mock(classOf[OffsetCheckpoints])
+
+    logManager.startup()
+    val partition = new Partition(topicPartition,
+      replicaLagTimeMaxMs = kafka.server.Defaults.ReplicaLagTimeMaxMs,
+      interBrokerProtocolVersion = ApiVersion.latestVersion,
+      localBrokerId = brokerId,
+      mockTime,
+      stateStore,
+      delayedOperations,
+      metadataCache,
+      logManager) {
+
+      override def needsExpandIsr(followerReplica: Replica): Boolean = {
+        if (isrCheckDelayMs > 0)
+          Thread.sleep(isrCheckDelayMs)
+        super.needsExpandIsr(followerReplica)
+      }
+
+      override def needsShrinkIsr(): Boolean = {
+        if (isrCheckDelayMs > 0)
+          Thread.sleep(isrCheckDelayMs)
+        super.needsShrinkIsr()
+      }
+
+      override def getOutOfSyncReplicas(maxLagMs: Long): Set[Int] = {
+        if (isrCheckDelayMs > 0)
+          Thread.sleep(isrCheckDelayMs)
+        super.getOutOfSyncReplicas(maxLagMs)
+      }
+
+      override def maybeUpdateIsrAndVersion(isr: Set[Int], zkVersionOpt: Option[Int]): Unit = {
+        if (isrUpdateDelayMs > 0)
+          Thread.sleep(isrUpdateDelayMs)
+        super.maybeUpdateIsrAndVersion(isr, zkVersionOpt)
+      }
+
+      override def createLog(replicaId: Int, isNew: Boolean, isFutureReplica: Boolean, offsetCheckpoints: OffsetCheckpoints): Log = {
+        val log = super.createLog(replicaId, isNew, isFutureReplica, offsetCheckpoints)
+        new SlowLog(log, mockTime, appendDelayMs)
+      }
+
+      override def inReadLock[T](lock: ReadWriteLock)(fun: => T): T = {
```

Comment:
This feels a bit low level to me. Presumably a different locking implementation could get the same benefit of avoiding contention. Do you think they add much value to these test case?

Reply:
Removed this and added semaphores for testing.